### PR TITLE
Add TypeScript declaration files

### LINKED ISSUE #91: TypeScript definitions for opentok-react
### Problem

`opentok-react` currently provides no TypeScript type definitions. This makes the package less attractive / usable for TypeScript codebases - consumers need to create their own type definitions or disable type checking for affected code.

### Proposed solution

Provide TypeScript definition files (`.d.ts`) with the library.

### PR DESCRIPTION
**Pull request by janjakubnanista**
Fixes #91 

TypeScript types were missing for `opentok-react` which is a pain for all TypeScript developers using the library at the moment. 

I added the .d.ts files for both `opentok-react` and the base `opentok` module. This is not ideal ofc since `opentok` types should not live in this project. **If there is a better place to put them please let me know!** It also makes this pull request very large :(

I also added `tsd` tests for the types I created, these will check the prop types of all exported components.

### CODE REVIEW

**Comment by nielsdB97** [comment]
After I started working on creating types for this package myself I found this PR. I think it looks great!
The only problem I found was that your typings don't allow for a `ref` to be passed to `OTPublisher` and `OTSubscriber`, whereas the README indicates you can (search for `ref={`), for which the best solution I found was to extend the props with `React.RefAttributes`:

```
export const OTPublisher: React.ComponentType<
    OTPublisherProps & React.RefAttributes<any>
  >;
```

**Comment by janjakubnanista** [comment]
@nielsdB97 I added the `ref` properties to both `OTSubscriberProps` and `OTPublisherProps` plus added the types for the global `OT` object

File: types/opentok.d.ts [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+  streamDestroyed?: EventHandler<StreamDestroyedEvent>;
+  streamPropertyChanged?: EventHandler<StreamPropertyChangedEvent>;
+} | Record<string, EventHandler<SignalEvent>>;
+
+/**
+ * Session object
+ * 
+ * {@link https://tokbox.com/developer/sdks/js/reference/Session.html}
+ */
+export interface Session extends EventEmitter<SessionEventHandlers> {
```
[the remark this answers is not in the document]
Comment (janjakubnanista) [reply types/opentok.d.ts:287]: I added the `isConnected` method to `Session`

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+export const OTSubscriber: React.ComponentType<OTSubscriberProps>;
+
+export interface CreateSessionOptions {
+  apiKey: string;
+  sessionId: string;
+  token: string;
+  onStreamsUpdated?: (streams: Stream[]) => void;
+}
+
+export interface CreateSessionResult {
```
[the remark this answers is not in the document]
Comment (janjakubnanista) [reply types/index.d.ts]: Aaaaand done!

File: types/opentok.d.ts [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+}
+
+/**
+ * Publisher object
+ * 
+ * {@link https://tokbox.com/developer/sdks/js/reference/Publisher.html}
+ */
+export interface Publisher extends EventEmitter<PublisherEventHandlers> {
+  accessAllowed: boolean;
+  element?: Element;
```
[the remark this answers is not in the document]
Comment (janjakubnanista) [reply types/opentok.d.ts]: Changed to `HTMLElement`

File: types/opentok.d.ts [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+  videoEnabled?: EventHandler<VideoEnabledChangedEvent<'videoEnabled'>>;
+};
+
+/**
+ * Subscriber object
+ * 
+ * {@link https://tokbox.com/developer/sdks/js/reference/Subscriber.html}
+ */
+export interface Subscriber extends EventEmitter<SubscriberEventHandlers> {
+  element: Element;
```
[the remark this answers is not in the document]
Comment (janjakubnanista) [reply types/opentok.d.ts]: Changed to `HTMLElement`

File: types/opentok.d.ts [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+  reason: 'publishVideo' | 'quality' | 'subscribeToVideo' | 'codecNotSupported' | 'codecChanged';
+}
+
+/**
+ * VideoElementCreatedEvent
+ * 
+ * {@link https://tokbox.com/developer/sdks/js/reference/VideoElementCreatedEvent.html}
+ */
+export interface VideoElementCreatedEvent extends Event<'videoElementCreated'> {
+  element: Element;
```
[the remark this answers is not in the document]
Comment (janjakubnanista) [reply types/opentok.d.ts]: Changed to `HTMLElement`

**Comment by janjakubnanista** [comment]
@nielsdB97 I see that the CI check is hanging, is that a known issue or does something need to be done?

**Comment by nielsdB97** [comment]
@janjakubnanista for some reason the Karma tests can't be run since travis is [failing to spawn a Chrome instance](https://travis-ci.org/github/opentok/opentok-react/jobs/684259206#L575) to run them against. According to [this comment](https://github.com/karma-runner/karma-chrome-launcher/issues/171#issuecomment-363456028) it might help to clear the Travis NPM cache, but it could also be that a new run will work, as I don't see any related code changes.

**Comment by mathieuseguin** [comment]
is there anything I could do to retry the CI build?

**Comment by ggoldens** [comment]
@msach22 should this PR be pointing to dev branch?

**Comment by msach22** [comment]
@janjakubnanista Thanks for the PR. Could you please open it against the dev branch?

**Comment by ggoldens** [comment]
Rebasing to dev.

**Comment by ggoldens** [comment]
@janjakubnanista thanks for your PR! I closed it and reopened since travis jobs were not working. Now I see it's not passing. Can you check it? Thanks!

**Comment by janjakubnanista** [comment]
> @janjakubnanista thanks for your PR! I closed it and reopened since travis jobs were not working. Now I see it's not passing. Can you check it? Thanks!

@ggoldens I see that karma is failing to capture the chrome browser for unit test, this looks like an unrelated issue at first sight (I remember issues like these with karma from a ling time ago but can't quite remember what it was). Does this issue maybe appear in other PRs?

**Comment by ggoldens** [comment]
> > @janjakubnanista thanks for your PR! I closed it and reopened since travis jobs were not working. Now I see it's not passing. Can you check it? Thanks!
> 
> @ggoldens I see that karma is failing to capture the chrome browser for unit test, this looks like an unrelated issue at first sight (I remember issues like these with karma from a ling time ago but can't quite remember what it was). Does this issue maybe appear in other PRs?

I opened another PR to fix travis: https://github.com/opentok/opentok-react/pull/135

**Comment by enricop89** [comment]
@janjakubnanista thanks for the PR. One quick question: shouldn't we add the `types` key on the package.json?

Something like: 

```
"types": "types/index.d.ts"
```

**Comment by janjakubnanista** [comment]
@enricop89 Yes indeed! Moreover there now is a new script in `package.json`, `test`, that can be used to make sure the type definitions don't break. I think it would make sense to take a look and put it somewhere in the pipeline.

**Comment by enricop89** [comment]
@janjakubnanista thanks, could you please push the changes? I'll review them

**Pull request merged**